File: 백준/Silver/2531. 회전 초밥/회전 초밥.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int d = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());

        int[] table = new int[n];
        int[] sushi = new int[d + 1];

        for(int i=0; i<n; i++){
            table[i] = Integer.parseInt(br.readLine());
        }

        sushi[c]++;
        int answer = 1;
        for(int i=0; i<k; i++){
            if(sushi[table[i]] == 0) answer++;
            sushi[table[i]]++;
        }

        int start = 0;
        int end = k-1;
        int count = answer;
        while(true){
            sushi[table[start]]--;
            if(sushi[table[start]] == 0) count--;

            if(start == n-1) start = 0;
            else start++;

            if(end == n-1) end = 0;
            else end++;

            if(end == k-1) break;
            sushi[table[end]]++;

            if(sushi[table[end]] == 1) count++;

            answer = Math.max(answer, count);
        }

        System.out.println(answer);
    }
}

File: 백준/Silver/2531. 회전 초밥/회전 초밥.py
N, d, k, c = map(int, input().split())
sushi = [int(input()) for _ in range(N)]

_set = set()
_set.add(c)

dict = {}
dict[c] = 1

for i in range(k):
    _set.add(sushi[i])
    dict[sushi[i]] = dict.get(sushi[i], 0) + 1

answer = len(_set)

start = 0
end = k-1
while start < N:
    dict[sushi[start]] = dict.get(sushi[start])-1

    if dict[sushi[start]] == 0:
        dict.pop(sushi[start], None)
        _set.remove(sushi[start])
    
    start += 1
    end = (end+1) % N

    dict[sushi[end]] = dict.get(sushi[end], 0) + 1
    _set.add(sushi[end])

    answer = max(answer, len(_set))


print(answer)